### DISCUSSION TITLE
did web method

### DISCUSSION THREAD

**Discussion by praveensankar** [decentralized-identity/veramo#388] [Q&A]
Is there any registry which stores all the web method dids and provides the resolution function to the agents ?

**Comment by sk91** [comment]
@praveensankar you are looking for something like [universal resolver](https://github.com/decentralized-identity/universal-resolver)?

**Reply by praveensankar** [reply]
I am looking for any existing verifiable data registries such as ethr-did-registry which acts as common storage (ledger or some database) for method type 'web'.

**Reply by sk91** [reply]
I don't believe so.

It will go against the principle of web-did method. Any website owner can publish their web-did on their website.

You can check if it exists by going into https://website.com/.well-known/did.json


for example:

for `identity.foundation` domain, to resolve the did you would go to

https://identity.foundation/.well-known/did.json

ethr-did does not track all did's as well by the way, only thous that need to add additional attributes on top of the default ones that can be derived by ethereum address

**Reply by praveensankar** [reply]
It basically boils down to DNS lookup operation. Is that correct ?

**Reply by sk91** [reply]
A http request:

https://w3c-ccg.github.io/did-method-web/#read-resolve

2.5.2 Read (Resolve)
The following steps MUST be executed to resolve the DID document from a Web DID:

- Replace ":" with "/" in the method specific identifier to obtain the fully qualified domain name and optional path.
- Generate an HTTPS URL to the expected location of the DID document by prepending https://.
- If no path has been specified in the URL, append /.well-known.
- Append /did.json to complete the URL.
- Perform an HTTP GET request to the URL using an agent that can successfully negotiate a secure HTTPS connection, which enforces the security requirements as described in § 2.6 Security and privacy considerations .

**Reply by praveensankar** [reply]
If I understand correctly for every DID lookup, we have to issue HTTP GET request which fetches the corresponding json file from the website. HTTP GET request will in turn rely on the DNS servers and caches to find the ip address of the domain and the corresponding resource. This whole DNS lookup delay is present here compared to directly fetching it from any standard systems such as blockchains or distributed databases.

**Reply by sk91** [reply]
yes, but can you describe your usecase that will make this delay meaningful for you? 

You can always cache the did documents for repeated resolution for reasonable timeframe. Did documents do not tend to change quickly

**Reply by praveensankar** [reply]
I am trying to understand the underlying working. I am not planning to applying it in any scenario yet. Thanks for the information.

**Comment by mirceanis** [comment]
@praveensankar In the case of `veramo` the agent can act as a registry for the `did:web` identifiers it manages.
Take a look at [WebDidDocRouter](https://github.com/uport-project/veramo/blob/57fa00393e5268e57c2b409952f3181f1f6a7a13/packages/remote-server/src/web-did-doc-router.ts#L28) in the `@veramo/remote-server` package.
It is an [expressjs](https://github.com/expressjs/express) router that can answer the `.well-known/did.json` request.
Note that it does not store JSON files, but builds the did.json response on the fly out of the known keys and services registered for the respective `did:web`

For quick experimentation, you can try to use the `@veramo/cli`. The current version of the CLI sets up a temporary `did:web` through ngrok when you use the `veramo server` command.
